nsis/src/nsis/NsisLoader.java

```diff
@@ -81,56 +91,82 @@ public String getName() {
+			throws CancelledException, IOException {
 		try {
 			GenericFactory factory = MessageLogContinuesFactory.create(log);
-			NsisExecutable ne = NsisExecutable.createNsisExecutable(factory,
-					provider, SectionLayout.FILE);
-
-			long nsis_header_offset = ne.getHeaderOffset();
-			if (nsis_header_offset == -1) {
-				System.out.print("Could not find nsis_header_offset.\n");
-				return;
+			NsisExecutable ne = NsisExecutable.createNsisExecutable(factory, provider,
+					SectionLayout.FILE);
+			long scriptHeaderOffset = ne.getHeaderOffset();
+			BinaryReader binary_reader = new BinaryReader(provider, /* isLittleEndian= */ true);
+			binary_reader.setPointerIndex(scriptHeaderOffset);
+			Address scriptHeaderAddress = program.getAddressFactory().getDefaultAddressSpace()
+					.getAddress(scriptHeaderOffset);
+
+			FileBytes fileBytesHeader = MemoryBlockUtils.createFileBytes(program, provider,
+					scriptHeaderOffset, NsisScriptHeader.getHeaderSize(), monitor);
+			initScriptHeader(fileBytesHeader, scriptHeaderAddress, fileBytesHeader.getSize(),
+					program, ne.getHeaderDataType());
+
+			FileBytes fileBytesBody;
+			byte[] decompressedBytes = ne.getDecompressedBytes();
+			if (decompressedBytes != null) { // if binary was compressed
+				ByteArrayProvider uncompressedBytes = new ByteArrayProvider(decompressedBytes);
+				fileBytesBody = MemoryBlockUtils.createFileBytes(program, uncompressedBytes, 0,
+						uncompressedBytes.length(), monitor);
+			} else {
+				fileBytesBody = MemoryBlockUtils.createFileBytes(program, provider,
+						scriptHeaderOffset + NsisScriptHeader.getHeaderSize(), ne.getArchiveSize(),
+						monitor);
 			}
-
-			// TODO we probably don't need both types of readers (input
-			// stream and binary reader)
-			InputStream inputStream;
-			inputStream = provider.getInputStream(nsis_header_offset);
-			Memory mem = program.getMemory();
-
-			BinaryReader binary_reader = new BinaryReader(provider,
-					/* isLittleEndian= */ true);
-			binary_reader.setPointerIndex(nsis_header_offset);
-
-			ghidra.program.model.address.Address script_header_start = program
-					.getAddressFactory().getDefaultAddressSpace()
-					.getAddress(0x0);
-
-			MemoryBlock new_block = mem.createInitializedBlock(".script_header",
-					script_header_start, inputStream,
-					ne.getInflatedHeaderSize(), monitor, false);
-			new_block.setRead(true);
-			new_block.setWrite(true);
-			new_block.setExecute(true);
-
-			createData(program, program.getListing(), script_header_start,
-					ne.getHeaderDataType());
-			processBlockHeaders(program, monitor, binary_reader,
-					nsis_header_offset);
-
-			System.out.printf("Done initializing block headers\n");
+			initBlockHeaders(program, binary_reader,
+					scriptHeaderAddress.add(NsisScriptHeader.getHeaderSize()), fileBytesBody);
 		} catch (Exception e) {
-			System.out.print(e.getMessage());
-			log.appendException(e);
+			throw new IOException(e); // Ghidra handles the thrown exception
```

Comment:
Because the function is a override of a function defined by Ghidra and it can only throw IOException and CancelledException, it can't throw our InvalidFormatException because that would modify the function's signature